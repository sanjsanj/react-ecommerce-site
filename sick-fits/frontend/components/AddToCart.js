import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
      # item {
      #   id
      # }
      # user {
      #   id
      # }
    }
  }
`;

class AddToCart extends Component {
  render() {
    const { id } = this.props;

    return (
      <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ id }}>
        {addToCart => <button onClick={addToCart}>Add to cart ðŸ›’</button>}
      </Mutation>
    );
  }
}

export default AddToCart;
