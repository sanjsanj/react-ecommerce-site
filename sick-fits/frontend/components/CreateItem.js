import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Router from "next/router";

import Form from "./styles/Form";

import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: "Default title",
    description: "Default description",
    image: "item.jpg",
    largeImage: "item-large.jpg",
    price: 2000
  };

  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === "number" ? parseFloat(value) : value;

    this.setState({
      [name]: val
    });
  };

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              e.preventDefault();
              const res = await createItem();
              
              Router.push({
                pathname: "/item",
                query: { id: res.data.createItem.id }
              })
            }}
          >
            <Error error={error} />

            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  value={this.state.title}
                  onChange={this.handleChange}
                  required
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  value={this.state.price}
                  onChange={this.handleChange}
                  required
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter a description"
                  value={this.state.description}
                  onChange={this.handleChange}
                  required
                />
              </label>

              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export { CREATE_ITEM_MUTATION };
export default CreateItem;
