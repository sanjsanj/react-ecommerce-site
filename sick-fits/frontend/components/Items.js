import React, { Component } from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";

import Pagination from "./Pagination";
import Item from "./Item";

const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY {
    items {
      id
      title
      description
      price
      image
      largeImage
    }
  }
`;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`;

class Items extends Component {
  render() {
    return (
      <Center>
        <Pagination page={this.props.page} />
          <Query query={ALL_ITEMS_QUERY}>
            {({ data, error, loading }) => {
              if (error) return <p>Error: {error.message}</p>;
              if (loading) return <p>Loading...</p>;
              return (
                <ItemsList>
                  {data.items.map(item => (
                    <Item key={item.id} item={item} />
                  ))}
                </ItemsList>
              );
            }}
          </Query>
        <Pagination page={this.props.page} />
      </Center>
    );
  }
}

export { ALL_ITEMS_QUERY };
export default Items;
