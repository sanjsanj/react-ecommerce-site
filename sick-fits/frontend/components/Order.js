import React, { Component } from "react";
import { adopt } from "react-adopt";
import PropTypes from "prop-types";
import { Query } from "react-apollo";
import { format } from "date-fns";
import Head from "next/head";
import gql from "graphql-tag";

import OrderStyles from "./styles/OrderStyles";

import formatMoney from "../lib/formatMoney";

import Error from "./ErrorMessage";

const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      id
      charge
      createdAt
      total
      user {
        id
      }
      items {
        id
        title
        description
        image
        largeImage
        price
        quantity
      }
    }
  }
`;

const Composed = adopt({
  orderQuery: ({ id, render }) => (
    <Query query={SINGLE_ORDER_QUERY} variables={{ id }}>
      {render}
    </Query>
  )
});

class Order extends Component {
  static proprTypes = {
    id: PropTypes.string.isRequired
  };

  render() {
    return (
      <Composed id={this.props.id}>
        {({ orderQuery }) => {
          if (orderQuery.error) return <Error error={orderQuery.error} />;

          if (orderQuery.loading) return <p>Loading...</p>;

          const order = orderQuery.data.order;

          return (
            <OrderStyles data-test="order">
              <Head>
                <title>Sick Fits - Order {order.id}</title>
              </Head>
              <p>
                <span>Order ID:</span>
                <span>{this.props.id}</span>
              </p>
              <p>
                <span>Charge</span>
                <span>{order.charge}</span>
              </p>
              <p>
                <span>Date</span>
                <span>{format(order.createdAt, "MMMM d, YYYY h:mm a")}</span>
              </p>
              <p>
                <span>Order Total</span>
                <span>{formatMoney(order.total)}</span>
              </p>
              <p>
                <span>Item Count</span>
                <span>{order.items.length}</span>
              </p>
              <div className="items">
                {order.items.map(item => (
                  <div className="order-item" key={item.id}>
                    <img src={item.image} alt={item.title} />
                    <div className="item-details">
                      <h2>{item.title}</h2>
                      <p>Qty: {item.quantity}</p>
                      <p>Each: {formatMoney(item.price)}</p>
                      <p>SubTotal: {formatMoney(item.price * item.quantity)}</p>
                      <p>{item.description}</p>
                    </div>
                  </div>
                ))}
              </div>
            </OrderStyles>
          );
        }}
      </Composed>
    );
  }
}

export default Order;
