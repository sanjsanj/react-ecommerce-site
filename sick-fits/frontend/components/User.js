import { Query } from "react-apollo";
import PropTypes from "prop-types";
import gql from "graphql-tag";

const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`;

const User = props => (
  <Query {...props} query={CURRENT_USER_QUERY}>
    {payload => props.children(payload)}
  </Query>
);

User.propTypes = {
  children: PropTypes.func.isRequired
};

export default User;
export { CURRENT_USER_QUERY };
