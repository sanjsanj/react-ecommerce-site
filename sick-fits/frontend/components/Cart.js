import React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";

import CartStyles from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import SickButton from "./styles/SickButton";

import { formatPossessive } from "../lib/formatPossessive";
import { formatPlural } from "../lib/formatPlural";
import calcTotalPrice from "../lib/calcTotalPrice";

import User from "./User";
import CartItem from "./CartItem";
import formatMoney from "../lib/formatMoney";

const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`;

const Cart = () => {
  return (
    <User>
      {({ data: { me } }) => {
        if (!me) return null;

        return (
          <Mutation mutation={TOGGLE_CART_MUTATION}>
            {toggleCart => (
              <Query query={LOCAL_STATE_QUERY}>
                {({ data }) => (
                  <CartStyles open={data.cartOpen}>
                    <header>
                      <CloseButton onClick={toggleCart} title="close">
                        &times;
                      </CloseButton>

                      <Supreme>{formatPossessive(me.name)} Cart</Supreme>

                      <p>
                        You have {me.cart.length}{" "}
                        {formatPlural(me.cart, "Item")} in your Cart
                      </p>
                    </header>

                    <ul>
                      {me.cart.map(cartItem => (
                        <CartItem key={cartItem.id} cartItem={cartItem} />
                      ))}
                    </ul>

                    <footer>
                      <p>{formatMoney(calcTotalPrice(me.cart))}</p>
                      <SickButton>Checkout</SickButton>
                    </footer>
                  </CartStyles>
                )}
              </Query>
            )}
          </Mutation>
        );
      }}
    </User>
  );
};

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };
