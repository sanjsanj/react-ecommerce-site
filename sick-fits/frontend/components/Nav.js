import Link from "next/link";

import NavStyles from "./styles/NavStyles";

import User from "./User";

const Nav = () => (
  <NavStyles>
    <User>
      {({ data: { me } }) => {
        return me ? <p>{me.name}</p> : null;
      }}
    </User>

    <Link href="/items">
      <a>Shop</a>
    </Link>

    <Link href="/sell">
      <a>Sell</a>
    </Link>

    <Link href="/signup">
      <a>Signup</a>
    </Link>

    <Link href="/orders">
      <a>Orders</a>
    </Link>

    <Link href="/me">
      <a>Me</a>
    </Link>
  </NavStyles>
);

export default Nav;
