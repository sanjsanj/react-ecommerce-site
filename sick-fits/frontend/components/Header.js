import Link from "next/link";

import Nav from "./Nav";

const Header = () => (
  <div>
    <div className="bar">
      <Link href="/">
        <a>Sick Fits</a>
      </Link>
    </div>

    <Nav />

    <div className="sub-bar">
      <p>Search</p>
    </div>

    <div>Cart</div>
  </div>
);

export default Header;
