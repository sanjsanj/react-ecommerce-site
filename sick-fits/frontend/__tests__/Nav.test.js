import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import waait from "waait";
import { MockedProvider } from "react-apollo/test-utils";

import { fakeUser, fakeCartItem } from "../lib/testUtils";

import { CURRENT_USER_QUERY } from "../components/User";
import Nav from "../components/Nav";

const notSignedInMocks = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: { data: { me: null } }
  }
];

const signedInMocks = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: { data: { me: fakeUser() } }
  }
];

const signedInMocksWithCartItems = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem()]
        }
      }
    }
  }
];

describe("<Nav />", () => {
  it("renders non signed in nav when not signed in", async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav />
      </MockedProvider>
    );

    await waait();
    wrapper.update();

    expect(wrapper.text()).toContain("Shop");
    expect(wrapper.text()).toContain("Signin");
    expect(wrapper.text()).toContain("Signup");

    const nav = wrapper.find("[data-test='nav']");

    expect(toJSON(nav)).toMatchSnapshot();
  });

  it("renders signed in nav when signed in", async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav />
      </MockedProvider>
    );

    await waait();
    wrapper.update();

    expect(wrapper.text()).toContain("Sell");
    expect(wrapper.text()).toContain("Orders");
    expect(wrapper.text()).toContain("Account");
    expect(wrapper.text()).toContain("My Cart");

    const nav = wrapper.find("[data-test='nav']");

    expect(nav.find("Link").length).toEqual(4);
  });

  it("renders right amount of cart items", async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocksWithCartItems}>
        <Nav />
      </MockedProvider>
    );

    await waait();
    wrapper.update();

    const nav = wrapper.find('[data-test="nav"]');
    const count = nav.find("div.count");
    
    expect(count.text()).toEqual("3");
    expect(toJSON(count)).toMatchSnapshot();
  });
});
