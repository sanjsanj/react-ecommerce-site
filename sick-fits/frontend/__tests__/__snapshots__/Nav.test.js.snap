// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders non signed in nav when not signed in 1`] = `
Array [
  <NavStyles
    data-test="nav"
  >
    <ul
      className="NavStyles-sc-11c0d2g-0 iBzfrx"
      data-test="nav"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
        >
          Shop
        </a>
      </Link>
      <Link
        href="/signup"
      >
        <a
          href="/signup"
          onClick={[Function]}
        >
          Signup
        </a>
      </Link>
      <Link
        href="/signup"
      >
        <a
          href="/signup"
          onClick={[Function]}
        >
          Signin
        </a>
      </Link>
    </ul>
  </NavStyles>,
  <ul
    className="NavStyles-sc-11c0d2g-0 iBzfrx"
    data-test="nav"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
      >
        Shop
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        Signup
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        Signin
      </a>
    </Link>
  </ul>,
]
`;

exports[`<Nav /> renders right amount of cart items 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 kDZyEW"
>
  3
</div>
`;
